### ISSUE TITLE
Version 1.3.2: Uploading image throws Network error on Android

### ISSUE THREAD

**Issue by thegitparticle** [GetStream/stream-chat-react-native#375]
After upgrading to 1.3.2, the image upload throws network error on Android. Every thing else works fine, internet connection is solid but this happening. 
This is probably due to the new image crop and picker library added, please check. Lot of RN image picker libraries throw this issue due to not able to set the file source rightly on Android. 

Let me know how to fix this in this version?

gz#6494

**Comment by shodgetts** [comment]
Agent comment from Seetha in <a href='https://getstream.zendesk.com/agent/tickets/6494'>Zendesk ticket #6494</a>:<br><br>Hi

Thanks for reaching out to us.
We will get this checked for you.

Cheers
Stream Team

**Comment by vishalnarkhede** [comment]
Hey @thegitparticle can you please tell us what error are you seeing exactly? Screenshot or screenrecording will be even better :)

**Comment by thegitparticle** [comment]
The image just does not upload. And console logs Network error. 

Attaching image below...

![Uploading Screenshot_20201007-153839_Aye.jpg…]()

**Comment by thegitparticle** [comment]
@vishalnarkhede any update?

**Comment by DanC5** [comment]
@thegitparticle I just built our current master branch to my android device and am not having any issue uploading an image. Can you upgrade to 1.3.3 and if you still have an issue send me a screenshot of the network error?

**Comment by thegitparticle** [comment]
Tried with 1.3.3 and 0.63.2 - still facing the issue with image and file uploads.

**Comment by vishalnarkhede** [comment]
Hey @thegitparticle Can you confirm I you have done all following steps:

1. Install react-native-image-crop-picker package in your app - `yarn add react-native-image-crop-picker`
2. `cd ios && pod install`
3. In Xcode open Info.plist and add string key NSPhotoLibraryUsageDescription with value that describes why you need access to user photos. More info here https://forums.developer.apple.com/thread/62229. Depending on what features you use, you also may need NSCameraUsageDescription and NSMicrophoneUsageDescription keys.

**Comment by thegitparticle** [comment]
@vishalnarkhede I have react-native-image-crop-picker installed already. Tried with 0.34.1 and 0.35.0 as well,  both versions throw the same issue, Also, the issue is on Android, haven't tried it yet on iOS, so maybe not the issue with pods.

**Comment by vishalnarkhede** [comment]
@thegitparticle in case of android, can you confirm that you have followed steps mentioned here - https://github.com/ivpusic/react-native-image-crop-picker#android

**Comment by thegitparticle** [comment]
Yes, I did check. All the things are intact. Yet, the error persists. 

I used to get similar error with react-native-image-picker by R-N-community a while back when using in a different component (unrelated to stream), and the solution for that was to upgrade my Flipper and embed file:/// to the path.


```
        let image = {uri: 'file:///' + response.path};
        setImageSource(image.uri);
```

Maybe see if the this is the issue

**Comment by thegitparticle** [comment]
@vishalnarkhede any update on this?

**Comment by vishalnarkhede** [comment]
`Hey @thegitparticle you can override our internal image picker functionality to try out the solution suggested by you:

```js
import ImagePicker from 'react-native-image-crop-picker';
import {registerNativeHandlers} from 'stream-chat-react-native-core';

registerNativeHandlers({
  pickImage: () => {
    try {
      let res = await ImagePicker.openPicker({
        includeBase64: Platform.OS === 'ios',
        maxFiles: 3 || undefined,
        multiple: true,
        writeTempFile: false,
      });

      return {
        cancelled: false,
        images: res.map((image) => ({
          uri:
            Platform.OS === 'ios'
              ? `data:${image.mime};base64,${image.data}`
              : image.path,
        })),
      };
    } catch (err) {
      return {
        cancelled: true,
      };
    }
  }
});
```

**Comment by thegitparticle** [comment]
where do I find this file?

**Comment by thegitparticle** [comment]
Is it stream-chat-react-native > src > index.js ??? Is so, when I try to console.log image or res, I didn't get logged anything when I try to upload (use this functionality). that logging could help me insert the right prefix for the file.

Also, if this doesn't solve, could you tell me a way to change image-crop-picker to the previous library you were using? I really can not delay my launch even further beyond today.

**Comment by vishalnarkhede** [comment]
This is how you can go to previous image picker functionality (haven't tested this ... but should work)

```js
import ImagePicker from 'react-native-image-picker';
import {registerNativeHandlers} from 'stream-chat-react-native-core';

registerNativeHandlers({
  pickImage: () =>
    new Promise((resolve, reject) => {
      ImagePicker.showImagePicker({}, (response) => {
        if (response.error) {
          reject(Error(response.error));
        }
        let { uri } = response;
        if (Platform.OS === 'android') {
          uri = 'file://' + response.path;
        }

        resolve({
          cancelled: response.didCancel,
          images: [{
              uri
          }],
        });
      });
    }),
});
```

**Comment by thegitparticle** [comment]
This change has to be done in here - [stream-chat-react-native > src > index.js] right?

**Comment by vishalnarkhede** [comment]
No. ... you can do it in your app. You just need to call `registerNativeHandlers` function before rendering your chat components.

**Comment by thegitparticle** [comment]
I did, just called the function as you said before the return (<Chat></Chat>), I am still getting Network error. 

Is there anything else I missed in switching the library?

**Comment by thegitparticle** [comment]
I created a new project and re implemented the whole thing. Yet, the error persists with both image-crop-picker and the old image-picker as well. 

But, this time, the source of error as it shows seem diff for image-picker, attaching screenshot, @vishalnarkhede have a look please. 

![Screenshot_20201009-090314_mock4](https://user-images.githubusercontent.com/12975157/95541141-316e5180-0a10-11eb-9bcd-5d7502c536d9.jpg)

**Comment by ivanph** [comment]
Hi everyone, I ran into this issue and I think I found the solution and cause.

**TLDR:  The `stream-chat` package v2.5.0 broke the format for the file object. Downgrade to a previous version.**

Same as OP all I could see was the `Network Error` in the app. I added a logger to the client to see if the issue was actually coming from the picker (which I was almost sure wasn't the case because the images did load in the app).

When looking at the `extraData` for the upload request error this caught my eye

`_response": "Binary FormData part needs a content-type header.",`

Looking at the form data content the files looked like this

```
"payload": {
    "_parts": [
      [
        "file",
        {
          "uri": "file:///data/user/0/com.nasm.edge.debug/cache/react-native-image-crop-picker/output.jpg",
          "name": "output.jpg",
          "contentType": "image/jpeg"
        }
      ]
    ]
  },
```
If you notice there is a `contentType` key that should actually be `type`.
After this I went through the source, trying to find were the file object is created, but didn't see anything obvious. 
At this point I decided to take a look at the `stream-chat` package to see if there were any recent changes related to sending images and right away this commit got my attention https://github.com/GetStream/stream-chat-js/commit/b62aac7a26e31ac17b29c12212344acfdd22d54e 
When the code that adds the file to the FormData object was moved to the `addFileToFormData` function, the `type` field was renamed to `contenType` and this is why the XMLHTTPRequest impl for react-native complains about the missing content-type header from this line https://github.com/facebook/react-native/blob/b581bcaae0fbc6a245856d8bb7631311253f10fc/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L710 (and seems to only happen on Android).
I was able to confirm this by downgrading `stream-chat` to 2.2.2 and also by just editing the files locally to change `contentType` to `type`

**Comment by thegitparticle** [comment]
@ivanph I will try this out.

I downgraded stream to 1.2.0 and it worked. But the issue now is png files are not being uploaded, because seems like the default image format is put as image/jpeg. I am attaching the error screenshot below, @vishalnarkhede could you guide me here?
![Screenshot 2020-10-09 at 4 00 03 PM](https://user-images.githubusercontent.com/12975157/95573046-7f9f4700-0a48-11eb-8880-6d07c59c0a98.png)

**Comment by ivanph** [comment]
@thegitparticle pngs are working for me.
 What I did was
```
npm i stream-chat-react-native@latest
npm i stream-chat@2.2.2
```

**Comment by virdesai** [comment]
@thegitparticle I've opened a PR fix for this with our JS client [here](https://github.com/GetStream/stream-chat-js/pull/464)

**Comment by thegitparticle** [comment]
@ivanph will try this.

**Comment by ivanph** [comment]
I can confirm that this is fixed after upgrading `stream-chat` to v2.6.0.
Thanks everyone.

**Comment by panda919** [comment]
I got network  error yet on get-stream-react-native 2.0.0-beta.0 yet on android .

![image](https://user-images.githubusercontent.com/25161426/96470016-70967100-1260-11eb-8364-78386e3d31e7.png)

**Comment by damithsynapsys** [comment]
using the latest version of stream,

and get the same error; 

```
    "stream-chat": "^2.8.0",
    "stream-chat-react-native": "^2.0.0",
```

![image](https://user-images.githubusercontent.com/57521888/100479630-cd653100-3114-11eb-907f-ba7cba4c5959.png)
![Screenshot 2020-11-28 at 00 58 44](https://user-images.githubusercontent.com/57521888/100479661-df46d400-3114-11eb-8bf0-81c1520be654.png)

**Comment by vishalnarkhede** [comment]
@damithsynapsys So as part of recent investigation, we noticed few caveats around file uploads:

- Flipper interferes with file uploads on android, so please make sure you have updated version of flipper - where they have fixed the network interceptors
- Chrome debugger also interferes with upload requests on android. So please try testing without chrome debugger.

Please let me know if that fixes issue for you.

**Comment by damithsynapsys** [comment]
I have updated to 0.69.0 and the file upload started to work.

**Comment by Marcoo09** [comment]
@damithsynapsys what library yo update to 0.69.0?

**Comment by Marcoo09** [comment]
I still have this error both on android and ios.
I tried upgrading flipper to 0.69.0 but not fix the error.
    "stream-chat-react-native": "3.0.0",
    "@react-native-community/cameraroll": "4.0.2",

Any suggestion? @vishalnarkhede 

Thanks in advance!

**Comment by vishalnarkhede** [comment]
@Marcoo09 Have you test without flipper? Also could you please try upgrading `stream-chat-react-native` to latest version - 3.7.2

**Comment by Marcoo09** [comment]
Thanks Vishal @vishalnarkhede for the quick response.

Now I installed the 3.7.2 and remove flipper from both, android and ios:

- On IOS its working now
- On android I continue with the network error
- On both I can send images as attachement or take a photo
<img width="372" alt="Screen Shot 2021-08-24 at 12 00 18" src="https://user-images.githubusercontent.com/18408823/130640693-3b8a4041-7ac8-4888-83e8-f25c9e4a8b89.png">

Any other suggestion to fix the selection directly from gallery?

**Comment by vishalnarkhede** [comment]
@marco-ulge are you running any kind of debugger e.g., chrome debugger or reactotron? Debuggers have well know issues around file uploads on android, since they add some network interceptors. So best to test in release mode just to confirm.
